File: culinator-desktop/src/features/recipe-editor/narrative.ts
import { computed, type ComputedRef, type Ref, unref } from "vue";
import {
  narrativeWasm,
  type WasmNumberStyle,
  type WasmUnitSystem,
} from "../../services/wasm/parser";

export { sortOperationsForDisplay } from "./operation-order";

/**
 * Recipe narrative for the reading page.
 *
 * Every sentence, ingredient line, time chip, and mise-en-place list here is
 * rendered by `culinator-narrative` (via WASM) — the same generator the EPUB,
 * print, and plain-text exporters use. This file used to derive all of it in
 * TypeScript, which meant the heuristics existed twice and drifted: the reading
 * page dropped step destinations ("Transfer the sauce mix." instead of "…into
 * the casserole") and rendered "0.5 tsp" where the exporter said "1/2 tsp".
 *
 * Amounts are converted and formatted in Rust too, so the page no longer makes
 * an async round-trip per quantity.
 */

/** An ingredient as displayed, split so a UI can column-align the amount. */
export interface IngredientDisplayParts {
  /** Resource symbol, for views that key rows by declaration. */
  symbol: string;
  /** "3 tbsp", "1/2", or empty when the amount is the cook's call. */
  amount: string;
  /** Size, state, and name: "large Hass avocados". */
  description: string;
  /** Trailing modifiers: divided, notes, to taste, optional. */
  aside?: string;
}

export interface IngredientGroup {
  /** Variant label, e.g. "Sweet" — omitted for the base ingredient list. */
  label?: string;
  items: IngredientDisplayParts[];
}

/** What one method section needs on hand before its steps start. */
export interface SectionMise {
  ingredients: IngredientDisplayParts[];
  equipment: string[];
}

export interface NarrativeStep {
  /** Operation symbol, so the editor can act on the right declaration. */
  symbol: string;
  number: number;
  /** Full instruction prose, equipment and doneness woven in. */
  text: string;
  time?: string;
  /** "hands-on · makes roux". */
  meta?: string;
  tools: string[];
}

export interface MethodSection {
  process: string;
  /** Humanized heading; absent when the recipe has a single process. */
  title?: string;
  /** Parallelism guidance for the section as a whole. */
  note?: string;
  steps: NarrativeStep[];
  mise: SectionMise;
}

export interface RecipeNarrative {
  summary: ComputedRef<string>;
  ingredientGroups: ComputedRef<IngredientGroup[]>;
  /** Whole-recipe equipment for the traditional top-matter layout. */
  equipment: ComputedRef<string[]>;
  sections: ComputedRef<MethodSection[]>;
  /** Flat step list, for layouts that number straight through. */
  steps: ComputedRef<NarrativeStep[]>;
}

export interface NarrativeOptions {
  unitSystem?: Ref<WasmUnitSystem> | ComputedRef<WasmUnitSystem>;
  numberStyle?: Ref<WasmNumberStyle> | ComputedRef<WasmNumberStyle>;
}

// The WASM boundary hands back plain JSON; these mirror its serde shape.
interface WasmLine {
  symbol: string;
  quantity: string;
  description: string;
  aside?: string;
}
interface WasmNarrative {
  summary: string;
  ingredientGroups: { label?: string; items: WasmLine[] }[];
  equipment: string[];
  sections: {
    process: string;
    title?: string;
    note?: string;
    steps: NarrativeStep[];
    mise: { ingredients: WasmLine[]; equipment: string[] };
  }[];
}

function toParts(line: WasmLine): IngredientDisplayParts {
  return {
    symbol: line.symbol,
    amount: line.quantity,
    description: line.description,
    aside: line.aside,
  };
}

/** Rendered ingredient lines keyed by resource symbol. */
export function ingredientPartsBySymbol(source: string): Map<string, IngredientDisplayParts> {
  return new Map(
    buildNarrative(source)
      .ingredientGroups.flatMap((group) => group.items)
      .map((item) => [item.symbol, item]),
  );
}

export interface BuiltNarrative {
  summary: string;
  ingredientGroups: IngredientGroup[];
  equipment: string[];
  sections: MethodSection[];
}

/** Render a recipe's narrative in one call. */
export function buildNarrative(
  source: string,
  unitSystem: WasmUnitSystem = "as_authored",
  numberStyle: WasmNumberStyle = "fractions",
): BuiltNarrative {
  const raw = narrativeWasm(source, unitSystem, numberStyle) as WasmNarrative;
  return {
    summary: raw.summary,
    ingredientGroups: raw.ingredientGroups.map((group) => ({
      label: group.label,
      items: group.items.map(toParts),
    })),
    equipment: raw.equipment,
    sections: raw.sections.map((section) => ({
      process: section.process,
      title: section.title,
      note: section.note,
      steps: section.steps,
      mise: {
        ingredients: section.mise.ingredients.map(toParts),
        equipment: section.mise.equipment,
      },
    })),
  };
}

export function useRecipeNarrative(
  source: Ref<string> | ComputedRef<string>,
  options: NarrativeOptions = {},
): RecipeNarrative {
  const narrative = computed(() =>
    buildNarrative(
      unref(source) ?? "",
      options.unitSystem ? unref(options.unitSystem) : "as_authored",
      options.numberStyle ? unref(options.numberStyle) : "fractions",
    ),
  );

  return {
    summary: computed(() => narrative.value.summary),
    ingredientGroups: computed(() => narrative.value.ingredientGroups),
    equipment: computed(() => narrative.value.equipment),
    sections: computed(() => narrative.value.sections),
    steps: computed(() => narrative.value.sections.flatMap((section) => section.steps)),
  };
}

/** First N method steps, for compact book previews. */
export function previewSteps(source: string, count = 4): string[] {
  return buildNarrative(source)
    .sections.flatMap((section) => section.steps)
    .slice(0, count)
    .map((step) => step.text);
}

/** Compact ingredient lines for book previews. */
export function previewIngredientParts(source: string, count = 5): IngredientDisplayParts[] {
  return buildNarrative(source)
    .ingredientGroups.flatMap((group) => group.items)
    .slice(0, count);
}

/** The recipe's "10 ingredients · 9 steps · ~2 h" line. */
export function narrativeSummary(source: string): string {
  return buildNarrative(source).summary;
}

/**
 * Durations in minutes → "1 h 20 min". Kept in TypeScript because the
 * scheduling views compute their own totals from live timer state rather than
 * from a parsed recipe.
 */
export function formatDuration(minutes: number): string {
  if (!minutes || minutes <= 0) return "";
  if (minutes < 1) return `${Math.round(minutes * 60)} sec`;
  const total = Math.round(minutes);
  const hours = Math.floor(total / 60);
  const mins = total % 60;
  if (hours && mins) return `${hours} h ${mins} min`;
  if (hours) return `${hours} h`;
  return `${mins} min`;
}

File: culinator-wasm/src/narrative.rs
//! Projection of the shared prose generator (`culinator-narrative`) onto the
//! shape the reading page renders.
//!
//! The desktop app used to derive this itself in `narrative.ts`, which meant
//! every sentence heuristic existed twice and drifted. Now the exporter and the
//! reading page render from the same generator, so a step reads identically in
//! the app, in the EPUB, and in the printed page.

use culinator_core::{Recipe, ResourceKind, UnitSystem};
use culinator_narrative as narrative;
use culinator_narrative::NumberStyle;
use serde::Serialize;

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiIngredientLine {
    pub symbol: String,
    /// Amount as displayed; empty when the amount is the cook's call.
    pub quantity: String,
    /// Size, state, and name: "large Hass avocados".
    pub description: String,
    /// Trailing modifiers, for a UI that renders them apart from the name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub aside: Option<String>,
}

impl From<&narrative::IngredientLine> for UiIngredientLine {
    fn from(line: &narrative::IngredientLine) -> Self {
        Self {
            symbol: line.symbol.clone(),
            quantity: line.quantity.clone(),
            description: line.description.clone(),
            aside: line.aside.clone(),
        }
    }
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiIngredientGroup {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub label: Option<String>,
    pub items: Vec<UiIngredientLine>,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiMise {
    pub ingredients: Vec<UiIngredientLine>,
    pub equipment: Vec<String>,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiStep {
    /// Operation symbol, so the editor can act on the right declaration.
    pub symbol: String,
    pub number: usize,
    /// Full instruction prose with equipment and doneness woven in.
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub time: Option<String>,
    /// "hands-on · makes roux", already joined.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub meta: Option<String>,
    pub tools: Vec<String>,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiSection {
    pub process: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    /// Parallelism guidance ("You can work on this while Prep is under way.").
    #[serde(skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
    pub steps: Vec<UiStep>,
    /// What this section alone needs on hand, for the colocated layout.
    pub mise: UiMise,
}

#[derive(Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct UiNarrative {
    /// "10 ingredients · 9 steps · ~2 h 17 min total"
    pub summary: String,
    pub ingredient_groups: Vec<UiIngredientGroup>,
    /// Whole-recipe equipment list for the traditional top-matter layout.
    pub equipment: Vec<String>,
    pub sections: Vec<UiSection>,
}

fn mise(recipe: &Recipe, process: &str, style: NumberStyle) -> UiMise {
    let derived = narrative::section_mise(recipe, process, style);
    UiMise {
        ingredients: derived
            .ingredients
            .iter()
            .map(UiIngredientLine::from)
            .collect(),
        equipment: derived.equipment,
    }
}

/// Build the reading-page narrative. `unit_system` restates every quantity in
/// metric or US customary; `None` keeps the amounts exactly as authored.
pub fn build(recipe: &Recipe, unit_system: Option<UnitSystem>, style: NumberStyle) -> UiNarrative {
    // Convert once, up front, so prose and mise agree on every amount.
    let converted;
    let recipe = match unit_system {
        Some(system) => {
            converted = narrative::convert_recipe_units(recipe, system);
            &converted
        }
        None => recipe,
    };
    let content = narrative::extract_with(recipe, style);

    // Ingredient groups come back as flat strings; re-derive the structured
    // lines so the UI can column-align amounts.
    let lines: Vec<narrative::IngredientLine> = recipe
        .resources
        .iter()
        .filter(|resource| resource.kind == ResourceKind::Ingredient)
        .map(|resource| narrative::ingredient_line_with(resource, None, style))
        .collect();
    let ingredient_groups = content
        .ingredient_groups
        .iter()
        .map(|group| UiIngredientGroup {
            label: group.label.clone(),
            items: group
                .items
                .iter()
                .filter_map(|flat| {
                    lines
                        .iter()
                        .find(|line| &line.flat() == flat)
                        .map(UiIngredientLine::from)
                })
                .collect(),
        })
        .collect();

    UiNarrative {
        summary: content.summary.clone(),
        ingredient_groups,
        equipment: content.equipment.clone(),
        sections: content
            .sections
            .iter()
            .map(|section| UiSection {
                process: section.process.clone(),
                title: section.title.clone(),
                note: section.note.clone(),
                steps: section
                    .steps
                    .iter()
                    .map(|step| UiStep {
                        symbol: step.symbol.clone(),
                        number: step.number,
                        text: step.text.clone(),
                        time: step.time.clone(),
                        meta: (!step.meta.is_empty()).then(|| step.meta.join(" · ")),
                        tools: step.tools.clone(),
                    })
                    .collect(),
                mise: mise(recipe, &section.process, style),
            })
            .collect(),
    }
}
